cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1873,24 +2089,18 @@ class ObjectOutputStream final : public io::OutputStream {
     state->completed_parts[slot] = std::move(part);
   }
 
-  static Status UploadPartError(const S3Model::UploadPartRequest& req,
-                                const S3Model::UploadPartOutcome& outcome) {
-    return ErrorToStatus(
-        std::forward_as_tuple("When uploading part for key '", req.GetKey(),
-                              "' in bucket '", req.GetBucket(), "': "),
-        "UploadPart", outcome.GetError());
-  }
-
  protected:
   std::shared_ptr<S3ClientHolder> holder_;
   const io::IOContext io_context_;
   const S3Path path_;
   const std::shared_ptr<const KeyValueMetadata> metadata_;
   const std::shared_ptr<const KeyValueMetadata> default_metadata_;
   const bool background_writes_;
+  const bool allow_delayed_open_;
 
   Aws::String upload_id_;
   bool closed_ = true;
+  bool is_multipart_created_ = false;
```

Comment:
For the record, `is_multipart_created_ == true` iff `!upload_id_.empty()`, right? Perhaps this can be consolidated and we can rename `upload_id_` to something more explicit, such as `multipart_upload_id`.